presto-main/src/main/java/com/facebook/presto/execution/QueryLimitMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,160 @@
+    private final MemoryPoolListener memoryPoolListener = this::onMemoryReserved;
+
+    @Inject
+    public QueryLimitMemoryRevokingScheduler(
+            LocalMemoryManager localMemoryManager,
+            SqlTaskManager sqlTaskManager,
+            TaskManagementExecutor taskManagementExecutor)
+    {
+        this(
+                ImmutableList.copyOf(getMemoryPools(localMemoryManager)),
+                requireNonNull(sqlTaskManager, "sqlTaskManager cannot be null")::getQueryContext,
+                requireNonNull(taskManagementExecutor, "taskManagementExecutor cannot be null").getExecutor());
+        log.debug("Using QueryLimitMemoryRevokingScheduler spilling strategy");
+    }
+
+    @VisibleForTesting
+    QueryLimitMemoryRevokingScheduler(
+            List<MemoryPool> memoryPools,
+            Function<QueryId, QueryContext> queryContextSupplier,
+            ScheduledExecutorService taskManagementExecutor)
+    {
+        this.memoryPools = ImmutableList.copyOf(requireNonNull(memoryPools, "memoryPools is null"));
+        this.queryContextSupplier = requireNonNull(queryContextSupplier, "queryContextSupplier is null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        registerPoolListeners();
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.removeListener(memoryPoolListener));
+    }
+
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool, QueryId queryId, long queryTotalMemoryUse)
+    {
+        try {
+            QueryContext queryContext = queryContextSupplier.apply(queryId);
+            long maxTotalMemory = queryContext.getMaxTotalMemory();
+            if (queryTotalMemoryUse <= maxTotalMemory) {
+                return;
+            }
+
+            log.debug("Scheduling check for %s", queryId);
+            scheduleRevoking(queryContext, maxTotalMemory);
+        }
+        catch (Throwable e) {
+            log.error(e, "Error when acting on memory pool reservation");
+        }
+    }
+
+    private void scheduleRevoking(QueryContext queryContext, long maxTotalMemory)
+    {
+        taskManagementExecutor.execute(() -> {
+            try {
+                revokeMemory(queryContext, maxTotalMemory);
+            }
+            catch (Throwable e) {
+                log.error(e, "Error requesting memory revoking");
+            }
+        });
+    }
+
+    @VisibleForTesting
+    synchronized void revokeMemory(QueryContext queryContext, long maxTotalMemory)
+    {
+        // get a fresh value for queryTotalMemory in case it's changed (e.g. by a previous revocation request)
+        long queryTotalMemory = getTotalQueryMemoryReservation(queryContext);
```

Comment:
Does this actually reflect in progress revocations bytes? I think this will return all the bytes that are already being revoked if I am reading `OperatorContext.requestMemoryRevoking` correctly.

I think the behavior we want is to fully check the entire condition from `onMemoryReserved` as if any in-flight revocations had already completed.

Reply:
The `synchronized` just guarantees that nothing else is requesting revocation at the same time as you are.  As for the requested revocations not being completed, I don't think it matters because we revoke in the same task order, so if the revocable memory is still being included, it would be included in both the total memory and the number of revoked bytes (possibly with some slight differences if some extra memory got actually revoked between when we calculated the total memory and called requestMemoryRevoke).  It would just mean that an operator can get requestMemoryRevoking multiple times.  It won't have any effect though. 

This is the same as for the other memory revoking strategies where you can have sequential memory reservations before the memory is actually revoked. Looking closer i'm actually sure what "checkPending" gets you for those other strategies. It doesn't even wait for the revocation loop to complete before setting checkPending to false.

Reply:
sorry, just double checked, and requestMemoryRevoke returns 0 bytes if memory revoking has already been requested.  I'm not sure if that makes sense.  Maybe I should change it to return the number of bytes still to be revoked, even if it's in process.  

It's still the case, that I think the other strategies have the exact same issue.

Reply:
So confusing :-) OK so it will do the revocation, but it's mostly a no-op unless previous revocations have completed  and it can trigger a new revocation. But if it triggers a new revocation it will probably be based on the number of bytes currently needing to be revoked so I think it's fine?

Also on this pass of review I am now noticing that it's revoking tasks based on task id instead of how many bytes would be freed up which is possibly less efficient since you could end up spilling things like joins with very small builds sides and then you need to spill the entire probe when it wasn't actually necessary.

Reply:
lol https://github.com/prestodb/presto/pull/15898#discussion_r606380709.  but also fair point about the joins with small build sides and giant probes. I'll switch to a tree map ordered by size.

Reply:
Heh. The actual optimal choice of what to spill is probably something like "prefer to spill aggregation unless there is a build side that is greater than some threshold to spill". But that is a wild guess and to do this optimally would be a project in and of itself.